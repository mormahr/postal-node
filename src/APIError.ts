export default class APIError extends Error {
  data: { [key: string]: any };

  constructor(data: { [key: string]: any }) {
    super('Error executing postal API request');
    this.data = data;
  }
}
