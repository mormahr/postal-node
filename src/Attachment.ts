export interface AttachmentData {
  content_type: string;
  data: string;
  name: string;
}

export function makeAttachment(
  filename: string,
  contentType: string,
  data: Buffer
): AttachmentData {
  return {
    content_type: contentType,
    data: data.toString('base64'),
    name: filename,
  };
}
