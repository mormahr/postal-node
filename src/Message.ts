/**
 * Basic info of message that is returned by the API if no other information is requested.
 * See {@link https://apiv1.postalserver.io/controllers/messages/message.html}
 */
export interface Message {
  id: number;
  token: string;
}

/**
 * "Expansions" offered by the API.
 */
export interface ExtendedMessage {
  status: {
    status: string; // Sent, ?
    last_delivery_attempt: number;
    held: boolean;
    hold_expiry: unknown;
  };
  details: {
    rcpt_to: string;
    mail_from: string;
    subject: string;
    message_id: string;
    timestamp: number;
    // TODO: test
    direction: 'outgoing' | 'incoming';
    size: string;
    bounce: number;
    bounce_for_id: number;
    tag: unknown;
    // TODO: test
    received_with_ssl: 1 | 0;
  };
  inspection: {
    inspected: boolean;
    spam: boolean;
    spam_score: number;
    threat: boolean;
    threat_details: unknown;
  };
  plain_body: string | undefined;
  html_body: string | undefined;
  attachments: { [key: string]: string };
  headers: { [key: string]: { [key: string]: string[] } };
  raw_message: string;
  activity_entries: {
    loads: unknown[];
    clicks: unknown[];
  };
}
