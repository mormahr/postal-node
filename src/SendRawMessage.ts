export interface SendRawMessage {
  mail_from: string;
  rcpt_to: string;
  data: Buffer;
}
