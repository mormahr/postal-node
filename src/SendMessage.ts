import { AttachmentData } from './Attachment';

/**
 * @link https://apiv1.postalserver.io/controllers/send/message.html
 */
export interface SendMessage {
  to: string[];
  cc?: string[];
  bcc?: string[];
  from: string;
  sender?: string;
  subject?: string;
  tag?: string;
  reply_to?: string;
  plain_body: string;
  html_body?: string;
  attachments?: AttachmentData[];
  headers?: { [key: string]: string };
  bounce?: boolean;
}
