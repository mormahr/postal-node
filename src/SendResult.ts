import { Message } from './Message';

export interface SendResult {
  recipients: { [key: string]: Message };
}
