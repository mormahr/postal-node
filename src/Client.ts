import { ExtendedMessage, Message } from './Message';

import APIError from './APIError';
import { SendMessage } from './SendMessage';
import { SendRawMessage } from './SendRawMessage';
import { SendResult } from './SendResult';
import { SendResultRaw } from './SendResultRaw';
import fetch from 'node-fetch';

export default class Client {
  host: string;
  serverKey: string;

  constructor(host: string, serverKey: string) {
    this.host = host;
    this.serverKey = serverKey;
  }

  private async makeRequest(
    method: 'GET' | 'POST',
    controller: string,
    action: string,
    parameters: { [key: string]: any }
  ): Promise<any> {
    const path = `/api/v1/${controller}/${action}`;
    const url = new URL(path, this.host);
    const body = JSON.stringify(parameters);

    const response = await fetch(url, {
      method,
      headers: {
        'Content-Type': 'application/json',
        'X-Server-API-Key': this.serverKey,
      },
      body,
    });

    const result = await response.json();
    if (result.status === 'success') {
      return result.data;
    } else {
      throw new APIError(result);
    }
  }

  private readMessageResult(result: SendResultRaw): SendResult {
    const recipients: { [key: string]: Message } = {};
    for (let key of Object.keys(result.messages)) {
      recipients[key.toLowerCase()] = result.messages[key];
    }

    return { recipients };
  }

  async send(message: SendMessage): Promise<SendResult> {
    const result = await this.makeRequest('POST', 'send', 'message', message);
    return this.readMessageResult(result);
  }

  async sendRaw(message: SendRawMessage): Promise<SendResult> {
    const result = await this.makeRequest('POST', 'send', 'message', message);
    return this.readMessageResult(result);
  }

  async inspect(id: number): Promise<Message & ExtendedMessage> {
    const result = await this.makeRequest('GET', 'messages', 'message', {
      id,
      _expansions: true,
    });
    return result;
  }
}
