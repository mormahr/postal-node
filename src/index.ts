import { AttachmentData } from './Attachment';
import Client from './Client';
import { Message } from './Message';
import { SendMessage } from './SendMessage';
import { SendResult } from './SendResult';
import { makeAttachment } from './Attachment';

export {
  Client,
  AttachmentData,
  makeAttachment,
  Message,
  SendMessage,
  SendResult,
};
