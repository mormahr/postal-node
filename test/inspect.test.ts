import 'fetch-mock-jest';

import { Client } from '../src';
import { FetchMockStatic } from 'fetch-mock';
import fetch from 'node-fetch';
import fs from 'fs/promises';

jest.mock('node-fetch', () => require('fetch-mock-jest').sandbox());

const fetchMock = (fetch as unknown) as FetchMockStatic;

describe('inspect', () => {
  beforeEach(() => {
    fetchMock.reset();
  });

  const client = new Client('https://example.com', 'some-key');

  it('inspects with all expansions', async () => {
    fetchMock.get('https://example.com/api/v1/messages/message', {
      body: (
        await fs.readFile(__dirname + '/mock-data/messages_message_full.json')
      ).toString('utf-8'),
    });

    const result = await client.inspect(0);
    expect(result).toMatchSnapshot();
    // To test that TypeScript correctly infers the result type
    expect(result.details.bounce).toBe(0);
    expect(fetchMock).toHaveLastFetched(
      'https://example.com/api/v1/messages/message',
      {
        body: { id: 0, _expansions: true },
        method: 'GET',
      }
    );
  });
});
