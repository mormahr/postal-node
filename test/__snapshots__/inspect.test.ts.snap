// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspect inspects with all expansions 1`] = `
Object {
  "activity_entries": Object {
    "clicks": Array [],
    "loads": Array [],
  },
  "attachments": Array [],
  "details": Object {
    "bounce": 0,
    "bounce_for_id": 0,
    "direction": "outgoing",
    "mail_from": "def@example.com",
    "message_id": "afb5a6fe-b978-46d1-833d-6b8798ea58e7@rp.mail.example.com",
    "rcpt_to": "abc@example.com",
    "received_with_ssl": 1,
    "size": "1508",
    "subject": "Some mail",
    "tag": null,
    "timestamp": 1633467930.482752,
  },
  "headers": Object {
    "cc": Array [
      "xxx@example.com",
    ],
    "content-transfer-encoding": Array [
      "7bit",
    ],
    "content-type": Array [
      "multipart/alternative; boundary=\\"--==_mimepart_615cbe1a69b74_482af3534567f0419a5\\"; charset=UTF-8",
    ],
    "date": Array [
      "Tue, 05 Oct 2021 21:05:30 +0000",
    ],
    "dkim-signature": Array [
      "[fake sig]",
    ],
    "from": Array [
      "Def <def@example.com>",
    ],
    "message-id": Array [
      "<afb5a6fe-b978-46d1-833d-6b8798ea58e7@rp.mail.example.com>",
    ],
    "mime-version": Array [
      "1.0",
    ],
    "received": Array [
      "from api (postal_nginx.traefik_proxy_network [192.168.112.53]) by Postal with HTTP; Tue, 05 Oct 2021 21:05:30 +0000",
    ],
    "subject": Array [
      "Some mail",
    ],
    "to": Array [
      "abc@example.com",
    ],
    "x-postal-msgid": Array [
      "erU4xQPhUrLb",
    ],
  },
  "html_body": "<b>Fake message</b>",
  "id": 961,
  "inspection": Object {
    "inspected": false,
    "spam": false,
    "spam_score": 0,
    "threat": false,
    "threat_details": null,
  },
  "plain_body": "Fake message",
  "raw_message": "[fake raw message]",
  "status": Object {
    "held": false,
    "hold_expiry": null,
    "last_delivery_attempt": 1633467933.337266,
    "status": "Sent",
  },
  "token": "erU4xQPhUrLb",
}
`;
